src/module-loader/dotnet/src/MorganStanley.ComposeUI.ModuleLoader/ModuleLoader.cs

```diff
@@ -0,0 +1,106 @@
+﻿// Morgan Stanley makes this available to you under the Apache License,
+// Version 2.0 (the "License"). You may obtain a copy of the License at
+// 
+//      http://www.apache.org/licenses/LICENSE-2.0.
+// 
+// See the NOTICE file distributed with this work for additional information
+// regarding copyright ownership. Unless required by applicable law or agreed
+// to in writing, software distributed under the License is distributed on an
+// "AS IS" BASIS, WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express
+// or implied. See the License for the specific language governing permissions
+// and limitations under the License.
+
+using System.Collections.Concurrent;
+using System.Reactive.Subjects;
+
+namespace MorganStanley.ComposeUI.ModuleLoader;
+
+internal sealed class ModuleLoader : IModuleLoader, IAsyncDisposable
+{
+    private readonly Subject<LifetimeEvent> _lifetimeEvents = new();
+    private readonly ConcurrentDictionary<Guid, IModuleInstance> _modules = new();
+    private readonly Dictionary<string, IModuleRunner> _moduleRunners;
+    private readonly IModuleCatalog _moduleCatalog;
+    private readonly IReadOnlyList<IStartupAction> _startupActions;
+
+    public ModuleLoader(
+        IModuleCatalog moduleCatalog,
+        IEnumerable<IModuleRunner> moduleRunners,
+        IEnumerable<IStartupAction> startupActions)
+    {
+        ArgumentNullException.ThrowIfNull(moduleCatalog);
+        ArgumentNullException.ThrowIfNull(moduleRunners);
+        ArgumentNullException.ThrowIfNull(startupActions);
+
+        _moduleCatalog = moduleCatalog;
+        _moduleRunners = moduleRunners.GroupBy(runner => runner.ModuleType).ToDictionary(g => g.Key, g => g.First());
+        _startupActions = new List<IStartupAction>(startupActions);
+    }
+
+    public IObservable<LifetimeEvent> LifetimeEvents => _lifetimeEvents;
+
+    public async Task<IModuleInstance> StartModule(StartRequest request)
+    {
+        var manifest = _moduleCatalog.GetManifest(request.ModuleId);
+        if (manifest == null)
+        {
+            throw new Exception($"Unknown Module id: {request.ModuleId}");
+        }
+
+        if (!_moduleRunners.TryGetValue(request.ModuleId, out var moduleRunner))
+        {
+            throw new Exception($"No module runner available for {manifest.ModuleType} module type");
+        }
+
+        Guid instanceId = Guid.NewGuid();
+        var moduleInstance = new ModuleInstance(instanceId, manifest, request);
+        _modules.TryAdd(instanceId, moduleInstance);
+
+        _lifetimeEvents.OnNext(new LifetimeEvent.Starting(moduleInstance));
+        var startupContext = new StartupContext(request);
+
+        int index = -1;
+        Func<Task> nextAction = null!;
+        nextAction = () =>
```

Comment:
This is a bit hazardous, if I call the same `next` multiple times from my startup action, I expect the same (next) startup action to be called. Maybe a unique delegate should be created for each startup action.
```cs
var pipeline = _startupActions
    .Reverse()
    .Aggregate(
        () => Task.CompletedTask, 
        (next, action) => () => action.InvokeAsync(startupContext, next));

await moduleRunner.Start(moduleInstance, startupContext, pipeline);
```

Reply:
Nice tip, thank you. Updated the code